### PR TITLE
New: `class-methods-use-this` rule (fixes #5139)

### LINKED ISSUE #5139: no pure functions inside of es6 classes
I'd like to write a rule that complains if a class method does not use this. Sort of the opposite of no-invalid-this.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30473627-no-pure-functions-inside-of-es6-classes?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by gyandeeps**
**What issue does this pull request address?**
Add  `class-methods-use-this` rule.

**What changes did you make? (Give an overview)**
Created a brand new rule

**Is there anything you'd like reviewers to focus on?**
* I have activated the rule for `eslint` repo since we have started using class syntax now.

Credit to Patrick Williams (@pwmckenna) who created the rule but was not able to finish the work on it. (#5152)

### CODE REVIEW

File: lib/rules/class-methods-use-this.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/**
+ * @fileoverview Rule to enforce that all class methods use 'this'.
+ * @author Patrick Williams
```
Comment (platinumazure) [inline lib/rules/class-methods-use-this.js]: c/p error?

File: tests/lib/rules/class-methods-use-this.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
+/**
+ * @fileoverview Tests for class-methods-use-this rule.
+ * @author Patrick Williams
+ * @copyright 2016 Patrick Williams. All rights reserved.
```
Comment (platinumazure) [inline tests/lib/rules/class-methods-use-this.js]: c/p error on \@author? And the copyright is definitely wrong, that should say jQuery foundation

File: tests/lib/rules/class-methods-use-this.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+ruleTester.run("class-methods-use-this", rule, {
+    valid: [
+        {code: "class A { constructor() {} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { foo() {this} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { foo() {this.bar = 'bar';} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { foo() {bar(this);} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { foo() { if(true) { return this; } } }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { static foo() {} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { constructor() {} }", parserOptions: { ecmaVersion: 6 }},
```
Comment (platinumazure) [inline tests/lib/rules/class-methods-use-this.js]: Duplicate of first valid test?

File: tests/lib/rules/class-methods-use-this.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        {code: "({ a(){} });", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { foo() { () => this; } }", parserOptions: { ecmaVersion: 6 }}
```
Comment (platinumazure) [inline tests/lib/rules/class-methods-use-this.js]: What about this? `{code: "({ a: function () {} });"}`? My assumption is that should be valid as well since it's not a class.

**Comment by platinumazure** [comment]
LGTM, except:

* [ ] Possible c/p errors in file headers, including incorrect copyright
* [ ] Suggested another test

File: lib/rules/class-methods-use-this.js (hunk 1 above)
Comment (ilyavolodin) [inline lib/rules/class-methods-use-this.js]: No copyright. We don't need copyrights inside rules, we have one copyright at the root, that covers the whole repository.

File: tests/lib/rules/class-methods-use-this.js (hunk 2 above)
Comment (ilyavolodin) [inline tests/lib/rules/class-methods-use-this.js]: Just remove copyright

File: lib/rules/class-methods-use-this.js (hunk 1 above)
Comment (platinumazure) [inline lib/rules/class-methods-use-this.js]: Awesome. There is a copyright in the tests though, which should be removed.
Thanks for clarifying!

File: docs/rules/class-methods-use-this.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+class A {
+    constructor() {
+        super(); // OK. constructor is exempt
+    }
+}
+
+class A {
+    static foo = function () {
```
Comment (mysticatea) [inline docs/rules/class-methods-use-this.md]: This is invalid syntax.

File: lib/rules/class-methods-use-this.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                stack[stack.length - 1] = true;
+            }
+        }
+
+        return {
+            FunctionDeclaration: enterFunction,
+            "FunctionDeclaration:exit": exitFunction,
+            FunctionExpression: enterFunction,
+            "FunctionExpression:exit": exitFunction,
+            ThisExpression: markThisUsed
```
Comment (mysticatea) [inline lib/rules/class-methods-use-this.js]: I think it needs `Super: markThisUse` also.

```js
class A extends B {
    foo() {
        super.foo()
    }
}
```

Comment (gyandeeps) [inline lib/rules/class-methods-use-this.js]: You are correct. Thanks.

**Comment by gyandeeps** [comment]
All the needed changes have been made.

**Comment by platinumazure** [comment]
LGTM, thanks for contributing! (And I apologize for not understanding that pwmckenna had started this.)

File: docs/rules/class-methods-use-this.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
+# Enforce that class methods utilize `this` (class-methods-use-this)
+
+If a class method does not use `this`, it can safely be made a static function.
+
+Its possible to have a class method which doesn't use `this`, such as:
```
Comment (btmills) [inline docs/rules/class-methods-use-this.md]: Grammar nit: missing apostrophe in "It's"

File: docs/rules/class-methods-use-this.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    sayHi() {
+        console.log("hi");
+    }
+}
+
+let aObj = new A();
+a.sayHi(); // => "hi"
+```
+
+Above example shows that `sayHi` method doesn't uses `this`, so we can easily make that as a static function, like this:
```
Comment (btmills) [inline docs/rules/class-methods-use-this.md]: I'd reword this a bit:

> In the example above, the `sayHi` method doesn't use `this`, so we can safely make it a static method:

File: docs/rules/class-methods-use-this.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+```
+
+## Rule Details
+
+This rule is aimed to flag class methods that do not use `this`.
+
+Examples of **incorrect** code for this rule:
+
+```js
+/*eslint class-methods-use-this: 2*/
```
Comment (btmills) [inline docs/rules/class-methods-use-this.md]: From a quick glance over the docs, it looks like we're using the string levels ("off"/"warn"/"error") in code examples now.

File: lib/rules/class-methods-use-this.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        /**
+         * Checks if we are leaving a function that is a method, and reports if 'this' has not been used.
+         * Static methods and the constructor are exempt.
+         * Then pops the context off the stack.
+         * @param {ASTNode} node - A function node that was entered.
+         * @returns {void}
+         * @private
+         */
+        function exitFunction(node) {
+            if (isInstanceMethod(node.parent) && !stack[stack.length - 1]) {
```
Comment (btmills) [inline lib/rules/class-methods-use-this.js]: If you find it cleaner, you could instead use the return value from `pop()`:

```js
function exitFunction(node) {
    const methodUsesThis = stack.pop();
    if (isInstanceMethod(node.parent) && !methodUsesThis) {
        // ...
    }
}

**Comment by btmills** [comment]
I had a few suggestions for the docs and one possible change to the code (if you like it), otherwise LGTM!

File: lib/rules/class-methods-use-this.js (hunk 10 above)
Comment (gyandeeps) [inline lib/rules/class-methods-use-this.js]: :+1:

File: docs/rules/class-methods-use-this.md (hunk 9 above)
Comment (gyandeeps) [inline docs/rules/class-methods-use-this.md]: Since i bought a old closed PR, thats why it still says 2. will fix.

File: docs/rules/class-methods-use-this.md (hunk 8 above)
Comment (gyandeeps) [inline docs/rules/class-methods-use-this.md]: :+1:

File: lib/rules/class-methods-use-this.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+         * Then pops the context off the stack.
+         * @param {ASTNode} node - A function node that was entered.
+         * @returns {void}
+         * @private
+         */
+        function exitFunction(node) {
+            if (isInstanceMethod(node.parent) && !stack.pop()) {
+                context.report(node, "Expected 'this' to be used by class method '" + node.parent.key.name + "'.");
+            }
+            stack.pop();
```
Comment (btmills) [inline lib/rules/class-methods-use-this.js]: Calling `pop()` on line 53 made this call redundant.

Comment (btmills) [inline lib/rules/class-methods-use-this.js]: I'm guessing Travis isn't failing because there are no tests that cover nested methods. Worth adding a test for that?

Comment (gyandeeps) [inline lib/rules/class-methods-use-this.js]: ah, nice catch. I almost forgot about the pop at the end.

**Comment by btmills** [comment]
LGTM, but I won't merge yet in case anyone else wants to give it one more pass.

**Comment by ilyavolodin** [comment]
LGTM, but waiting another day for others to look

File: docs/rules/class-methods-use-this.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+/*eslint-env es6*/
+class A {
+    foo() {
+        this.bar = "Hello World"; // OK, this is used
+    }
+}
+
+class A {
+    constructor() {
+        super(); // OK. constructor is exempt
```
Comment (mysticatea) [inline docs/rules/class-methods-use-this.md]: Sorry, this is invalid syntax.
If a class doesn't have `extends` clause, SuperCall throws a SyntaxError.
Also, this function is a correct code whether this is a constructor or not, since this has a `super`.

Comment (gyandeeps) [inline docs/rules/class-methods-use-this.md]: i will add extends clause.

Comment (mysticatea) [inline docs/rules/class-methods-use-this.md]: Or we can remove this SuperCall.

```
    constructor() {
        // OK. constructor is exempt
    }
```

**Comment by mysticatea** [comment]
LGTM!

**Comment by vitorbal** [comment]
Thanks, @gyandeeps! Awesome work!

**Comment by gyandeeps** [comment]
@vitorbal credits to @pwmckenna as he did all the work. i just reopened his PR.

**Comment by vitorbal** [comment]
Good point :) thanks a lot for your contribution, @pwmckenna!

**Comment by pwmckenna** [comment]
@gyandeeps thanks for picking up the torch! much appreciated.

**Comment by jpickwell** [comment]
I know I'm bumping a PR that's been merged for sometime, but I feel that this rule was misnamed.

Class methods are methods defined on the class itself, thus static. Instance methods are defined on an instance of a class. Javascript is prototypal, so you could also call instance methods prototype methods.

An aside, `this` can be used in class (a.k.a. static) methods.

```javascript
class SomeClass {
  static build() {
    return new this() // instead of new SomeClass()
  }
}
```
